server-spi/src/main/java/org/keycloak/storage/client/ClientLookupProvider.java

```diff
@@ -28,7 +28,66 @@
  * @version $Revision: 1 $
  */
 public interface ClientLookupProvider {
-    ClientModel getClientById(String id, RealmModel realm);
-    ClientModel getClientByClientId(String clientId, RealmModel realm);
-    List<ClientModel> searchClientsByClientId(String clientId, Integer firstResult, Integer maxResults, RealmModel realm);
+    
+    /**
+     * Exact search for a client by its internal ID.
+     * @param realm Realm to limit the search.
+     * @param id Internal ID
+     * @return Model of the client, or {@code null} if no client is found.
+     */
+    ClientModel getClientById(RealmModel realm, String id);
+
+    /**
+     * Exact search for a client by its internal ID.
+     * @param realm Realm to limit the search.
+     * @param id Internal ID
+     * @return Model of the client, or {@code null} if no client is found.
+     * @deprecated Use {@link #getClientById(org.keycloak.models.RealmModel, java.lang.String)} instead.
+     */
+    default ClientModel getClientById(String id, RealmModel realm) { return getClientById(realm, id); }
+
+    /**
+     * Exact search for a client by its public client identifier.
+     * @param realm Realm to limit the search for clients.
+     * @param clientId String that identifies the client to the external parties.
+     *   Maps to {@code client_id} in OIDC or {@code entityID} in SAML.
+     * @return Model of the client, or {@code null} if no client is found.
+     */
+    ClientModel getClientByClientId(RealmModel realm, String clientId);
+
+    /**
+     * Exact search for a client by its public client identifier.
+     * @param realm Realm to limit the search.
+     * @param clientId String that identifies the client to the external parties.
+     *   Maps to {@code client_id} in OIDC or {@code entityID} in SAML.
+     * @return Model of the client, or {@code null} if no client is found.
+     * @deprecated Use {@link #getClientByClientId(org.keycloak.models.RealmModel, java.lang.String)} instead.
+     */
+    default ClientModel getClientByClientId(String clientId, RealmModel realm) { return getClientByClientId(realm, clientId); }
+
+    /**
+     * Case-insensitive search for clients that contain the given string in their public client identifier.
+     * @param realm Realm to limit the search for clients.
+     * @param clientId Searched substring of the public client
+     *   identifier ({@code client_id} in OIDC or {@code entityID} in SAML.)
+     * @param firstResult First result to return. Ignored if negative or {@code null}.
+     * @param maxResults Maximim number of results to return. Ignored if negative or {@code null}.
+     * @return Model of the client, or {@code null} if no client is found.
+     * @deprecated Use {@link #searchClientsByClientId} instead.
```

Comment:
This one is actually not deprecated, right?

Reply:
Indeed, thanks for spotting this!